firestore/integration_test_internal/src/query_test.cc

```diff
@@ -236,11 +236,18 @@ TEST_F(QueryTest, TestKeyOrderIsDescendingForDescendingInequality) {
                   {"e", {{"foo", FieldValue::Double(21.0)}}},
                   {"f", {{"foo", FieldValue::Integer(66)}}},
                   {"g", {{"foo", FieldValue::Double(66.0)}}}});
-  QuerySnapshot snapshot = ReadDocuments(
+  const Query query =
```

Comment:
@ehsannas Here is the testing code we'd like your opinion on. This PR updates each query test to _also_ run a count query.

My thought was that this isn't the right approach because (a) this reduces the focus of these tests on query features, (b) adds coverage for various query features that should be tested on the underlying SDKs, and (c) does not scale to additional aggregations like SUM and AVERAGE.

@tom-andersen Argues that (a) this is great test coverage to have, (b) does not need to be replicated for SUM and AVERAGE, and (c) the integration tests are already not very focussed so the possible reduction in the tests' "focus" is worth it. (please correct me if I've mis-represented your position, Tom).

Reply:
For purpose of discussion, I think there are 3 different issues at play here:

1. Should the wrapper SDK do a thorough integration test, or simply verify it correctly wraps underlying SDK and defer to the underlying SDK to do a thorough test? A lot of tests (not just for count) could be simplified or removed if we defer to underlying SDK. 

2. Currently the C++ SDK has **_integration_** tests, in the sense that it requires making real calls to emulator/backend to function, thereby exercising many layers of code simultaneously. I think what we really want is **_unit_** tests that can test units of code in isolation by mocking out dependencies. I think this is fundamental to our disagreement, in the sense that I think Denver is applying advice normally given to unit tests that can have a narrow focus, when in reality these are integration tests that exercise the full stack.

3. Contrast between Unit Testing and Integration Testing:
[Unit Testing](https://engdoc.corp.google.com/eng/doc/devguide/testing/unit/overview.md?polyglot=cpp&cl=head):
- Focused
- Understandable
- Fast
- Deterministic
- Resilient
- Hermetic
[Integration Testing](https://engdoc.corp.google.com/eng/doc/devguide/testing/integration/what-is-integration.md?cl=head):
- Not Focused and/or Not Hermetic

Here are some of my thoughts:
1. Integration tests tell a story of multiple interactions (create data, query, then delete, makes sure is deleted, etc). When considering query operators, a part of that story is what documents are returned and does the count match.
2. If an integration test fails, there are many layers that could be at fault (lack of focus). A good integration test will give hints to isolate problem. If the count is wrong, the next natural question will be if the returned documents are wrong as well. If both are wrong, maybe there is an error in how we construct request to backend (since they share logic for serializing filter operators).
3. Our test suite is slow (mainly because it is written as integration tests). Breaking tests up into smaller tests will mean more time is spent on test setup/teardown of populating documents and other tasks. This adds up, and we should course correct. Some of this might be mitigated by tests sharing setup/teardown environment, but that could also lead to interfering tests.